Refactoring ProductViewModelService for readability and maintenance

Leonardo-Ferreira opened the pull request:
Resolves (No Fix)
Type: Refactor

## Issue
ProductViewModelService is a big class, core to the system, and it's methods were very long and heavily indented. Long methods are traditionally a nightmare to unit test and deeply-indented code is hard to read and check the flow.

## Solution
I refactored the code, basically extracting methods and inverting `IF` statements in order to make the code more readable by reducing indentation. This build on [PR 667](https://github.com/grandnode/grandnode/pull/667).

## Breaking changes
NONE

## Testing
Basic navigation, like PR 667

Leonardo-Ferreira committed:
Merge pull request #1 from grandnode/develop

Rebasing fork

Leonardo-Ferreira committed:
Refactored and extracted methods
Extracted methods for PreparePriceModel and PreparePictureModel
Refactored the price model preparing focusing on grouped products
Refactored code to make it more readable

Leonardo-Ferreira committed:
Refactor for readability
Refactored code aiming readability:
-Extracted methods
-Invert IF's thus removing identation
-Paralelized a couple of async operations

Leonardo-Ferreira committed:
Refactor - Method Extraction
The products detail method on the service was dismembered in several methods for better readbility and testability

Leonardo-Ferreira committed:
Refactor for code readbility
More refactor focusing on making code more readble... inverted some if's and decreased identation in general

Leonardo-Ferreira committed:
Merge pull request #2 from grandnode/develop

Updating Fork

Leonardo-Ferreira committed:
Implemented ICacheManager.TryGetValueAsync
Extended the ICacheManager interface, creating a async version for TryGetValue(key)
Implemented the method on the relevant classes and changed the implementation on the CacheExtensions

Leonardo-Ferreira committed:
Started Redis Cache Imp Replacement
Replaced Packages for Original StackExchange.Redis, update the base implementation and interfaces.
Refactored IDistributedRedisCacheExtended into ICacheManager
Changed the registration of cache

Leonardo-Ferreira committed:
Changed usage of GET to GetAsync

Leonardo-Ferreira commented:
@PatrykPorabik Hey man! would you mind reviewing this PR? I got a few more changes lined up, which builds upon this PR....
tsk for your attention!

Leonardo-Ferreira committed:
Refactor for better readbility

KrzysztofPajak commented:
@Leonardo-Ferreira 
About refactoring of class ProductViewModelService we want change the approach by creating small functions. For temporary we must remove your changes with Task.ContinueWith because we have some problems with capturing exceptions. It's definitly harder to find the error. 
We are really appreciate your contribution to the project!

Changes with cache we will accept after when you remove file ProductViewModelService.cs from your pull request.
Thank you

Leonardo-Ferreira committed:
Enabling cache of minemized jsons
Minemized jsons where default values such as null for objects, 0 (zero) for ints, "False" for booleans are not serialized, thus saving space on the cache and reducing the bandwidth necessary to operate the software.
Since this was not something that i wanted to test the whole thing over, i created a small adapatation on the GET

Leonardo-Ferreira committed:
Cache of Featured Products
Implemented the cache of the featured products. Performance impact is quite significant.
This cache has been set with a fixed duration of 15 minutes.
At a later commit i will parametize this

Leonardo-Ferreira commented:
hey @KrzysztofPajak tsk for the update!
I ended up creating quite a mess here so I closed this PR, refactored some changes into smaller PRs...
I was hoping you could join me at a slack I created ([here is the invite link](https://join.slack.com/t/grandnodeworkspace/shared_invite/enQtODU1MDIyNTEzMzE1LTllMWUyMDFhOTdiMTBkMGI4N2UwYjMxNWEwZDg4YjRjMTIxZjk5MzlhZWRmMzgyNTQ4NWExYWJlMzAzODgwMzU)) so I could bounce some ideas that I have regarding performance in general .
